### PR TITLE
JS/PY/RB: get ReDoSUtil in sync for ruby

### PR DESCRIPTION
**Pull request by erik-krogh**

### CODE REVIEW

**Comment by esbena** [comment]
From the PR title, it sounds to me like this only changes the Ruby implementation, but that is clearly not the case. Can you elaborate a bit?

**Comment by nickrolfe** [comment]
When I added the ReDoS queries for Ruby, I started from the existing Python parser and the JS/Python `ReDoSUtil.qll`, but I found the only forms of character class they handled were simple escapes like `\d` and `\w`. I had to extend them to support POSIX character classes like `[[:digit:]]` and the similar [`\p{}` construct](https://ruby-doc.org/core-3.0.2/Regexp.html#class-Regexp-label-Character+Properties).

I haven't reviewed @erik-krogh's changes closely yet, but it looks he's generalised the concept of a character class so that the code can be shared with Ruby without losing support for those constructs. Is that right?

**Comment by erik-krogh** [comment]
> I haven't reviewed @erik-krogh's changes closely yet, but it looks he's generalised the concept of a character class so that the code can be shared with Ruby without losing support for those constructs. Is that right?

That is exactly right.  
I introduced a new predicate `isEscapeClass(RegExpTerm term, string clazz)`, which is used instead of `RegExpCharacterClassEscape`.   

The new predicate generalizes escape classes, and "normalizes" them to `\d`, `\s`, `\w`.  
Each language must implement this new predicate.  
The JS/Python implementations are currently trivial, but the Ruby implementation maps e.g. `[[:digit:]]` to `\d`.

File: ruby/ql/test/query-tests/security/cwe-1333-exponential-redos/tst.rb [hunk 1]
```diff
@@ -369,4 +369,7 @@
 bad87 = /^X(\u0061|a)*Y$/
 
 # GOOD
-good43 = /^X(\u0061|b)+Y$/
\ No newline at end of file
+good43 = /^X(\u0061|b)+Y$/
+
+# NOT GODD
```
Comment (nickrolfe) [inline ruby/ql/test/query-tests/security/cwe-1333-exponential-redos/tst.rb:374]:
```suggestion
# NOT GOOD
```

**Review by nickrolfe — APPROVED** [review]
Just a typo in a comment, but otherwise looks great. Thanks for doing this!

**Review by nickrolfe — APPROVED** [review]

**Review by esbena — APPROVED** [review]

**Pull request merged**